### PR TITLE
EQL: Add ? param support

### PR DESCRIPTION
**Pull request by rw-access**
Relates to #51756 (closed with API support)

Add support for `?` params to EQL. This hasn't been integrated/tested with for the API, but that can be added here easily if this is the right PR for that.

### CODE REVIEW

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+public class ParameterTests extends ESTestCase {
+
+    private final EqlParser parser = new EqlParser();
+
+    public Expression expr(String source) {
+        return parser.createExpression(source);
+    }
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java]: Newline here, please.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        assertEquals(expr("a == ?", Collections.singletonList(true)), expr("a == true"));
+        assertEquals(expr("a == ?", Collections.singletonList(false)), expr("a == false"));
+        assertEquals(expr("a == ?", Collections.singletonList(null)), expr("a == null"));
+    }
+
+    public void testInvalidParameters() {
+        ParsingException ex = expectThrows(ParsingException.class,
+            () -> new EqlParser().createExpression("a + ?", Collections.singletonList(new Object())));
+        assertThat(ex.getMessage(), containsString("Invalid parameter"));
+    }
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java]: Newline

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+public class ParameterTests extends ESTestCase {
+
+    private final EqlParser parser = new EqlParser();
+
+    public Expression expr(String source) {
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java]: Is this method used somewhere else? If not, maybe make it `private`?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import org.elasticsearch.xpack.ql.expression.predicate.logical.Or;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.arithmetic.Neg;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.GreaterThan;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.GreaterThanOrEqual;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.LessThan;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.LessThanOrEqual;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.NotEquals;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.sql.proto.SqlTypedParamValue;
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java]: Lots of unused imports.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        assertThat(ex.getMessage(), containsString("Invalid parameter"));
+    }
+    public void testMultipleParameters() {
+        Expression expression = expr("(? + ? * ?) - ?", Arrays.asList(1, 2, 3, 4));
+        assertEquals(expression, expr("(1 + 2 * 3) - 4"));
+    }
+
+    public void testNotEnoughParameters() {
+        ParsingException ex = expectThrows(ParsingException.class,
+            () -> new EqlParser().createExpression("(? + ? * ?) - ?", Arrays.asList(1, 2, 3)));
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java]: Why creating a new EqlParser if there is already one at class level?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        assertEquals(expr("a == ?", Collections.singletonList(1)), expr("a == 1"));
+        assertEquals(expr("a == ?", Collections.singletonList(1.0)), expr("a == 1.0"));
+        assertEquals(expr("a == ?", Collections.singletonList(true)), expr("a == true"));
+        assertEquals(expr("a == ?", Collections.singletonList(false)), expr("a == false"));
+        assertEquals(expr("a == ?", Collections.singletonList(null)), expr("a == null"));
+    }
+
+    public void testInvalidParameters() {
+        ParsingException ex = expectThrows(ParsingException.class,
+            () -> new EqlParser().createExpression("a + ?", Collections.singletonList(new Object())));
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/parser/ParameterTests.java]: Why creating a new EqlParser if there is one at class level?

**Review by astefan — COMMENTED** [review]
I've left few comments.

Also, since `?` has other meaning in the grammar, as well, I would like to see a test that contains both uses of `?`. For example:

```
assertEquals(expr("a == ?\"String with literal 'slash' \\ characters included\" and b == ? or c == ?", Arrays.asList("test", false)),
                expr("a == ?\"String with literal 'slash' \\ characters included\" and b == 'test' or c == false"));
```

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java [hunk 7]
```diff
@@ -44,10 +46,19 @@
 import org.elasticsearch.xpack.ql.util.StringUtils;
 
 import java.util.List;
+import java.util.Map;
 
 
 public class ExpressionBuilder extends IdentifierBuilder {
 
+    protected final ParserParams params;
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java]: Why protected and not private?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java [hunk 8]
```diff
@@ -230,6 +241,34 @@ public Literal visitNullLiteral(EqlBaseParser.NullLiteralContext ctx) {
...
+
+    @Override
+    public Literal visitParamLiteral(ParamLiteralContext ctx) {
+        Object value = param(ctx.PARAM());
+        Source source = source(ctx);
+        DataType dataType = null;
+
+        try {
+            dataType = DataTypes.fromJava(value);
+        } catch (QlIllegalArgumentException ignored) {};
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java]: The try/catch is not needed.
And we never catch exceptions and ignore them.

**Review by costin — CHANGES_REQUESTED** [review]
Left some comments - more tests are needed to check that not only the param get replaced but also that tricky strings are properly added in the query.

**Comment by matriv** [comment]
@elastic/es-ql

**Comment by matriv** [comment]
After searching around regarding query params in SQL DBs, I saw that `?` is the one used for prepared statements, and then each DB uses different syntax for its procedural language (defining custom functions/stored procedures, etc) like: `@param`, `$param` or `:param`. In my opinion I would say let's proceed with the basic usage of `?` as a place holder and a list (array) of value replacements that correspond to those params.

**Pull request closed without merging**